open_lm/attention.py

```diff
@@ -0,0 +1,143 @@
+    # xformers assumes q, k, v are [batch, seq_len, heads, embed_dim]
+    # We assume that queries match the last part of the key / value sequences
+    # see (https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.fmha.attn_bias.LowerTriangularFromBottomRightMask)
+    # we would like to replace the mask generation with: mask = xops.fmha.attn_bias.LowerTriangularFromBottomRightMask()
+    # sadly we cannot us this because it needs xformers>=0.0.23 and this is not compatible with torch<2.1.1 while llm-foundry requires torch<2.1.1
+
+    mask = None
+    # If queries have shape [batch, 1, heads, dim] it means there is only one query in the sequence.
+    # In this case, there is no notion of causal masking, so we can just set the mask to None.
+    # This is actually needed to get the desired behavior with seq_len=1.
+    if is_causal and queries.shape[1] == keys.shape[1]:
+        mask = xops.LowerTriangularMask()
+    elif is_causal and queries.shape[1] > 1:
+        # Build causal mask that assumes queries are in the end of the sequence.
+        batch, q_seq_len, heads, _ = queries.shape
+        k_seq_len = keys.shape[1]
+        mask = get_rectangular_mask((batch, heads), q_seq_len, k_seq_len, queries.device, queries.dtype)
+    return xops.memory_efficient_attention(queries, keys, values, attn_bias=mask)
+
+
+def torch_attn(queries, keys, values, is_causal):
+    # Need to call contiguous in torch >=2.1, otherwise later calls to .view() fail.
+    # Possibly related: https://github.com/pytorch/pytorch/issues/110213 - behavior of scaled_dot_product_attention
+    # changed between 2.0 and 2.1
+    if is_causal and keys.shape[1] > queries.shape[1] > 1:
+        q_seq_len = queries.shape[1]
+        k_seq_len = keys.shape[1]
+        # Same as above, we would like to use:
+        # mask = xops.fmha.attn_bias.LowerTriangularFromBottomRightMask().materialize((1, 1, q_seq_len, k_seq_len), queries.dtype, queries.device)
+        mask = get_rectangular_mask((1, 1), q_seq_len, k_seq_len, queries.device, queries.dtype)
+        return (
+            F.scaled_dot_product_attention(
+                queries.transpose(1, 2), keys.transpose(1, 2), values.transpose(1, 2), attn_mask=mask
+            )
+            .transpose(1, 2)
+            .contiguous()
+        )
+    elif queries.shape[1] == 1:
+        return (
+            F.scaled_dot_product_attention(queries.transpose(1, 2), keys.transpose(1, 2), values.transpose(1, 2))
+            .transpose(1, 2)
+            .contiguous()
+        )
+    else:
+        return (
+            F.scaled_dot_product_attention(
+                queries.transpose(1, 2), keys.transpose(1, 2), values.transpose(1, 2), is_causal=is_causal
+            )
+            .transpose(1, 2)
+            .contiguous()
+        )
+
+
+ATTN_NON_LINEARITIES = {
+    "relu": F.relu,
+    "relu_squared": lambda x: torch.pow(F.relu(x), 2),
+    # "gelu": F.gelu, # goes to NaN with bais so comment out for now
```

Comment:
Not a big deal, but could explicitly just 0 out the illegal attn weights instead of adding -inf.